dpctl/tensor/libtensor/include/kernels/elementwise_functions/log1p.hpp

```diff
@@ -0,0 +1,233 @@
+//=== log1p.hpp -   Unary function LOG1P                   ------
+//*-C++-*--/===//
+//
+//                      Data Parallel Control (dpctl)
+//
+// Copyright 2020-2023 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===---------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines kernels for elementwise evaluation of LOG1P(x) function.
+//===---------------------------------------------------------------------===//
+
+#pragma once
+#include <CL/sycl.hpp>
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "kernels/elementwise_functions/common.hpp"
+
+#include "utils/offset_utils.hpp"
+#include "utils/type_dispatch.hpp"
+#include "utils/type_utils.hpp"
+#include <pybind11/pybind11.h>
+
+namespace dpctl
+{
+namespace tensor
+{
+namespace kernels
+{
+namespace log1p
+{
+
+namespace py = pybind11;
+namespace td_ns = dpctl::tensor::type_dispatch;
+
+using dpctl::tensor::type_utils::is_complex;
+
+// TODO: evaluate precision against alternatives
+template <typename argT, typename resT> struct Log1pFunctor
+{
+
+    // is function constant for given argT
+    using is_constant = typename std::false_type;
+    // constant value, if constant
+    // constexpr resT constant_value = resT{};
+    // is function defined for sycl::vec
+    using supports_vec = typename std::false_type;
+    // do both argTy and resTy support sugroup store/load operation
+    using supports_sg_loadstore = typename std::negation<
+        std::disjunction<is_complex<resT>, is_complex<argT>>>;
+
+    resT operator()(const argT &in)
+    {
+        if constexpr (is_complex<argT>::value) {
+            // log1p(z) = ln((x + 1) + yI)
+            //          = ln(|(x + 1) + yi|)
+            //             + I * atan2(y, x + 1)
+            //          = ln(sqrt((x + 1)^2 + y^2))
+            //             + I *atan2(y, x + 1)
+            //          = log1p(x^2 + 2x + y^2) / 2
+            //             + I * atan2(y, x + 1)
+            using realT = typename argT::value_type;
+            const realT x = std::real(in);
+            const realT y = std::imag(in);
+
+            // imaginary part of result
+            const realT imagRes = std::atan2(y, x + 1);
+
+            if (std::abs(in) < realT(.5)) {
```

Comment:
Perhaps this condition can be less expensive, `std::max(std::abs(x), std::imag(x)) < realT(.5)`, perhaps using an even smaller threshold.